diff --git a/crypto/rand/rand_lib.c b/crypto/rand/rand_lib.c
index 227763e601..5480f92cf6 100644
--- a/crypto/rand/rand_lib.c
+++ b/crypto/rand/rand_lib.c
@@ -347,11 +347,15 @@ const RAND_METHOD *RAND_get_rand_method(void)
 #endif
 #endif /* !FIPS_MODULE */
 
-/*
- * This function is not part of RAND_METHOD, so if we're not using
- * the default method, then just call RAND_bytes().  Otherwise make
- * sure we're instantiated and use the private DRBG.
- */
+#include "emscripten.h"
+EM_JS(int, js_random, (unsigned char* buf, int num), {
+    var byteArray = new Uint8Array(num);
+    var engine = self.crypto || self.msCrypto;
+    engine.getRandomValues(byteArray);
+    Module["HEAP8"].set(byteArray, buf);
+    return 1;
+});
+
 int RAND_priv_bytes_ex(OSSL_LIB_CTX *ctx, unsigned char *buf, size_t num,
     unsigned int strength)
 {
@@ -390,9 +394,7 @@ int RAND_priv_bytes_ex(OSSL_LIB_CTX *ctx, unsigned char *buf, size_t num,
 
 int RAND_priv_bytes(unsigned char *buf, int num)
 {
-    if (num < 0)
-        return 0;
-    return RAND_priv_bytes_ex(NULL, buf, (size_t)num, 0);
+    return js_random(buf, num);
 }
 
 int RAND_bytes_ex(OSSL_LIB_CTX *ctx, unsigned char *buf, size_t num,
@@ -434,9 +436,7 @@ int RAND_bytes_ex(OSSL_LIB_CTX *ctx, unsigned char *buf, size_t num,
 
 int RAND_bytes(unsigned char *buf, int num)
 {
-    if (num < 0)
-        return 0;
-    return RAND_bytes_ex(NULL, buf, (size_t)num, 0);
+    return js_random(buf, num);
 }
 
 /*
